os/arch/arm/src/amebasmart/amebasmart_mipi.c

```diff
@@ -0,0 +1,355 @@
+#include <tinyara/arch.h>
+#include <tinyara/semaphore.h>
+#include <tinyara/mipidsi/mipi_dsi.h>
+#include <tinyara/mipidsi/mipi_display.h>
+#include <arch/board/board.h>
+
+#include "chip.h"
+#include "PinNames.h"
+
+#undef mipierr
+#undef mipiinfo
+#if defined(CONFIG_DEBUG_MIPI)
+#define mipiinfo(format, ...)   printf(format, ##__VA_ARGS__)
+#define mipierr(format, ...)    printf(format, ##__VA_ARGS__)
+#else
+#define mipiinfo(format, ...)
+#endif
+#define mipierr(format, ...)   printf(format, ##__VA_ARGS__)
+
+static u32 vo_freq;
+static u32 send_cmd_done = 0;
+
+struct amebasmart_mipi_dsi_host_s {
+	struct mipi_dsi_host dsi_host;
+	MIPI_TypeDef *MIPIx;
+	MIPI_InitTypeDef *MIPI_InitStruct;
+};
+
+struct irq {
+	u32 num;
+	u32 data;
+	u32 priority;
+};
+
+/* Helpers */
+static void amebasmart_set_clock(void);
+static void amebasmart_check_freq(struct lcd_data *data);
+static void amebasmart_mipi_init_helper(FAR struct amebasmart_mipi_dsi_host_s *priv);
+static void amebasmart_register_interrupt(void);
+
+/* MIPI methods */
+static void amebasmart_mipidsi_send_cmd(MIPI_TypeDef *MIPIx, u8 cmd, u8 payload_len, const u8 *para_list);
+static int amebasmart_mipi_attach(FAR struct mipi_dsi_host *dsi_host, FAR struct mipi_dsi_device *dsi_device);
+static int amebasmart_mipi_detach(FAR struct mipi_dsi_host *dsi_host, FAR struct mipi_dsi_device *dsi_device);
+static int amebasmart_mipi_transfer(FAR struct mipi_dsi_host *dsi_host, FAR const struct mipi_dsi_msg *msg);
+
+/************************************************************************************
+ * Private Data
+ ************************************************************************************/
+/*initialize the MIPI IRQ info*/
+struct irq mipi_irq_info = {
+	.num = MIPI_DSI_IRQ,
+	.data = (u32) MIPI,
+	.priority = INT_PRI_MIDDLE,
+};
+
+static const struct mipi_dsi_host_ops g_dsi_ops = {
+	.attach = amebasmart_mipi_attach,
+	.detach = amebasmart_mipi_detach,
+	.transfer = amebasmart_mipi_transfer,
+};
+
+MIPI_InitTypeDef MIPI_InitStruct_g;
+
+static struct amebasmart_mipi_dsi_host_s g_dsi_host = {
+	.dsi_host = {
+		&g_dsi_ops
+	},
+	.MIPIx = MIPI,
+	.MIPI_InitStruct = &MIPI_InitStruct_g,
+};
+
+/***********Helpers ***********/
+
+static void amebasmart_mipi_fill_buffer(FAR struct mipi_dsi_host *dsi_host)
+{
+	FAR struct amebasmart_mipi_dsi_host_s *priv = (FAR struct amebasmart_mipi_dsi_host_s *)dsi_host;
+	u32 vtotal, htotal_bits, bit_per_pixel, overhead_cycles, overhead_bits, total_bits;
+	priv->MIPI_InitStruct->MIPI_VideoDataFormat = MIPI_VIDEO_DATA_FORMAT_RGB565;		//changed for st7785
+	switch (priv->MIPI_InitStruct->MIPI_VideoDataFormat) {
+	case MIPI_VIDEO_DATA_FORMAT_RGB565:
+		bit_per_pixel = 16;
+		break;
+	case MIPI_VIDEO_DATA_FORMAT_RGB666_PACKED:
+		bit_per_pixel = 18;
+		break;
+	case MIPI_VIDEO_DATA_FORMAT_RGB666_LOOSELY:
+	case MIPI_VIDEO_DATA_FORMAT_RGB888:
+	default:
+		bit_per_pixel = 24;
+		break;
+	}
+	priv->MIPI_InitStruct->MIPI_LaneNum = 1;			//changed to 1 for st7785
```

Comment:
Any reason changing lane num from 2 to 1 here?

Reply:
yes, lcd st7785 supports 1 data lane as given in datasheet. 
MIPI Dis play Serial Interface ( DSI V1.01 r11 and D PHY V1.0, 1 clock and 1 data lane pairs)

It doesn't work with 2